lms/static/scripts/frontend_apps/components/LMSFilePicker.tsx

```diff
@@ -184,18 +184,53 @@ export default function LMSFilePicker({
     }
   };
 
+  // Does the actual file fetching for the active directory, either by loading
+  // it from its children, or by going to the API to get them
+  const loadFilesToDisplay = useCallback(
+    (isReload = false): Promise<File[]> => {
+      const getNextAPICallInfo = () =>
+        folderPath[folderPath.length - 1]?.contents || listFilesApi;
+      const loadFilesFromApi = (): Promise<File[]> =>
+        apiCall({
+          authToken,
+          path: getNextAPICallInfo().path,
+        });
+
+      if (!isReload) {
+        // If the last element in current folderPath contains children, we use
+        // them without calling the API. Otherwise, we fetch files from the API
+        const children = folderPath[folderPath.length - 1]?.children;
+        return children ? Promise.resolve(children) : loadFilesFromApi();
+      }
+
+      return loadFilesFromApi().then(loadedFiles => {
+        const [, ...activePathWithoutRoot] = folderPath;
+
+        // Try to load the corresponding children from loaded files, if present,
+        // in case we are not in root folder but the whole file tree was loaded.
+        return activePathWithoutRoot.reduce((loadedFiles, folder) => {
```

Comment:
Even with the expanded commenting I am still dubious about this `reduce`: to wit, I had to go look up documentation on the corner-case handling of what gets returned if the array is empty and `initialValue` is provided. It's complicated by the context itself being complicated, with some APIs returning a tree of everything, and some just the files for the current folder.

It's unglamorous, but you could handle it with a boring ol' `for` loop if you were so inclined (this snippet assumes `async` conversion). It's clunky code but I think something like this could be a little easier for a future dev to navigate. Feel free to use another approach...I just think the cognitive load of the `reduce` is challenging.

```
      const loadedFiles = await loadFilesFromApi();

      const [, ...activePathWithoutRoot] = folderPath;

      let filesForFolder = loadedFiles;

      for (const folder of activePathWithoutRoot) {
        const folderFound = loadedFiles.find(
          fileOrFolder => fileOrFolder.id === folder.id
        );
        if (folderFound && folderFound.children) {
          filesForFolder = folderFound.children;
          break;
        }
      }
      
      return filesForFolder;
```

Reply:
> I had to go look up documentation on the corner-case handling of what gets returned if the array is empty and `initialValue`

I had to check it myself as well, so it's obviously not clear enough by itself.

I can only agree the `for` loop makes it way more obvious at one sight. I'll switch to that approach and see if maybe it still needs a couple of comments.